apps/phttpd/phttpd.c

```diff
@@ -0,0 +1,996 @@
+		snprintf(path, sizeof(path), "%s/%s%d", dir, LEVELDBFILE, i);
+		snprintf(mpath, sizeof(mpath), "%s/%s%d", dir, LEVELDBMEMFILE, i);
+		status = leveldb::DB::Open(options, path, mpath, &db->leveldb);
+		if (!status.ok()) {
+			D("error to open leveldb %s", path);
+			return -1;
+		}
+		D("done Open LevelDB dbfile %s memfile %s", path, mpath);
+
+		leveldb::WriteOptions write_options;
+		write_options.sync = false;
+		status = db->leveldb->Put(write_options, "100", "test");
+		if (!status.ok()) {
+			D("leveldb write error");
+		}
+		status = db->leveldb->Get(leveldb::ReadOptions(), "100", &val);
+		if (!status.ok()) {
+			D("leveldb read error");
+		}
+		status = db->leveldb->Delete(leveldb::WriteOptions(), "100");
+		if (!status.ok()) {
+			D("leveldb write error");
+		}
+		D("leveldb test done (error reported if any)");
+	}
+#endif
+#ifdef WITH_BPLUS
+	/* need B+tree ? */
+	if (db->flags & DF_BPLUS) {
+		int rc;
+		snprintf(path, sizeof(path), "%s/%s%d", dir, BPLUSFILE, i);
+		rc = btree_create_btree(path, ((gfile_t **)&db->vp));
+		D("btree_create_btree() done (%d) %s", rc, path);
+		if (rc != 0)
+			return -1;
+		else if (db->flags & DF_PASTE)
+			return 0;
+	}
+#endif /* WITH_BPLUS */
+	snprintf(path, sizeof(path), "%s/%s%d", dir, DATAFILE, i);
+	fd = open(path, O_RDWR | O_CREAT, S_IRWXU);
+	if (fd < 0) {
+		perror("open");
+		return -1;
+	}
+	if (db->flags & DF_MMAP) {
+		if (fallocate(fd, 0, 0, db->size) < 0) {
+			perror("fallocate");
+			close(fd);
+			return -1;
+		}
+		db->paddr = (char *)do_mmap(fd, db->size);
+		if (db->paddr == NULL) {
+			close(fd);
+			return -1;
+		}
+	}
+	db->fd = fd;
+	return 0;
+}
+
+static int
+phttpd_thread(struct nm_targ *targ)
+{
+	struct nm_garg *nmg = targ->g;
+	struct phttpd_global *g =
+		(struct phttpd_global *)nmg->garg_private;
+
+	if (init_db((struct dbctx *)targ->opaque, targ->me, g->dba.dir,
+		    g->dba.flags, g->dba.size / nmg->nthreads)) {
+		D("error on init_db");
+		return ENOMEM;
+	}
+	return 0;
+}
+
+void
+clean_dir(char *dirpath)
+{
+	DIR *dp;
+	struct dirent *ent;
+
+	if (!dirpath)
+		return;
+	if ((dp = opendir(dirpath)) == NULL) {
+		return;
+	}
+	while ((ent = readdir(dp))) {
+		char fullp[256]; // XXX
+		size_t l;
+
+		if (ent->d_name[0] == '.')
+			continue;
+		else if (strstr(ent->d_name, EXTMEMFILE) == NULL &&
+			 strstr(ent->d_name, BPLUSFILE) == NULL &&
+			 strstr(ent->d_name, DATAFILE) == NULL)
+			continue;
+		strncat(strncpy(fullp, dirpath, sizeof(fullp) - 2), "/", 2);
+		l = strlen(fullp) + strlen(ent->d_name) + 1;
+		if (l < sizeof(fullp)) {
+			strncat(fullp, ent->d_name, l);
+		}
+		//strncat(fullp, ent->d_name, sizeof(fullp) - strlen(fullp) - 1);
+		D("removing %s", fullp);
+		if (unlink(fullp))
+			perror("unlink");
+	}
+}
+
+char *
+chkpath(char *dir, char *f, size_t size, int rm)
+{
+	int fd, mode = O_RDWR|O_CREAT;
+	char *path = calloc(1, strlen(dir) + strlen(f) + 2);
+
+	if (path == NULL)
+		return NULL;
+	snprintf(path, sizeof(path), "%s/%s", dir, f);
```

Comment:
`path` is a pointer here, so `sizeof(path)` is just the size of a pointer. What you want is `strlen(dir)+strlen(f)+2`.